lld/MachO/ObjC.cpp

```diff
@@ -320,3 +342,1007 @@ void objc::checkCategories() {
+        return registerError(
+            "Failed to create symbol reference to non-deleted category");
+
+      catListIsecOffset += target->wordSize;
+    }
+  }
+  return true;
+}
+
+// This fully erases the merged categories, including their body, their names,
+// their method/protocol/prop lists and the __objc_catlist entries that link to
+// them.
+bool ObjcCategoryMerger::eraseMergedCategories() {
+  // We expect there to be many categories in an input __objc_catList, so we
+  // can't just, of which we will merge only some. Because of this, we can't
+  // just erase the entire __objc_catList, we need to erase the merged
+  // categories only. To do this, we generate a new __objc_catList and copy over
+  // all the un-merged categories and erase all the affected (and only the
+  // affected) __objc_catList's
+
+  // Map of InputSection to a set of offsets of the categories that were merged
+  std::map<ConcatInputSection *, std::set<uint64_t>> catListToErasedOffsets;
+
+  for (auto &mapEntry : categoryMap) {
+    for (InfoInputCategory &catInfo : mapEntry.second) {
+      if (!catInfo.wasMerged) {
+        continue;
+      }
+      catInfo.catListIsec->linkerOptimizeReason =
+          LinkerOptReason::CategoryMerging;
+      catListToErasedOffsets[catInfo.catListIsec].insert(
+          catInfo.offCatListIsec);
+    }
+  }
+
+  // If there were categories that we did not erase, we need to generate a new
+  // __objc_catList that contains only the un-merged categories, and get rid of
+  // the references to the ones we merged.
+  if (!generateCatListForNonErasedCategories(catListToErasedOffsets))
+    return false;
+
+  // Erase the old method lists & names of the categories that were merged
+  for (auto &mapEntry : categoryMap) {
+    for (InfoInputCategory &catInfo : mapEntry.second) {
+      if (!catInfo.wasMerged)
+        continue;
+
+      catInfo.catBodyIsec->linkerOptimizeReason =
+          LinkerOptReason::CategoryMerging;
+      tryEraseDefinedAtIsecOffset(catInfo.catBodyIsec, catLayout.nameOffset,
+                                  /*stringOnly=*/true);
+      tryEraseDefinedAtIsecOffset(catInfo.catBodyIsec,
+                                  catLayout.instanceMethodsOffset);
+      tryEraseDefinedAtIsecOffset(catInfo.catBodyIsec,
+                                  catLayout.classMethodsOffset);
+      tryEraseDefinedAtIsecOffset(catInfo.catBodyIsec,
+                                  catLayout.protocolsOffset);
+      tryEraseDefinedAtIsecOffset(catInfo.catBodyIsec,
+                                  catLayout.classPropsOffset);
+      tryEraseDefinedAtIsecOffset(catInfo.catBodyIsec,
+                                  catLayout.instancePropsOffset);
+    }
+  }
+
+  return true;
+}
+
+bool ObjcCategoryMerger::doMerge() {
+  if (!collectAndValidateCategoriesData())
+    return false;
+
+  for (auto &entry : categoryMap) {
+    // Can't merge a single category into the base class just yet.
+    if (entry.second.size() <= 1)
```

Comment:
Again, I think if a single entry case is removed in `categoryMap` inside `collectAndValidateCategoriesData`, this check is not needed.

Reply:
Resolving due to conclusion [here](https://github.com/llvm/llvm-project/pull/82928/files#r1507077779)